src/main/java/graphql/validation/rules/VariableTypesMatch.java

```diff
@@ -60,17 +61,18 @@ public void checkVariable(VariableReference variableReference) {
             return;
         }
         GraphQLInputType expectedType = getValidationContext().getInputType();
-        Optional<InputValueWithState> schemaDefault = Optional.ofNullable(getValidationContext().getArgument()).map(v -> v.getArgumentDefaultValue());
-        Value schemaDefaultValue = null;
-        if (schemaDefault.isPresent() && schemaDefault.get().isLiteral()) {
-            schemaDefaultValue = (Value) schemaDefault.get().getValue();
-        } else if (schemaDefault.isPresent() && schemaDefault.get().isSet()) {
-            schemaDefaultValue = ValuesResolver.valueToLiteral(schemaDefault.get(), expectedType, getValidationContext().getGraphQLContext(), getValidationContext().getI18n().getLocale());
-        }
         if (expectedType == null) {
-            // we must have a unknown variable say to not have a known type
+            // we must have an unknown variable say to not have a known type
             return;
         }
+        GraphQLArgument argument = getValidationContext().getArgument();
+        Optional<InputValueWithState> schemaDefault = Optional.ofNullable(argument).map(v -> v.getArgumentDefaultValue());
+        Value<?> schemaDefaultValue = null;
+        if (schemaDefault.isPresent() && schemaDefault.get().isLiteral()) {
+            schemaDefaultValue = (Value<?>) schemaDefault.get().getValue();
+        } else if (schemaDefault.isPresent() && schemaDefault.get().isSet()) {
+            schemaDefaultValue = ValuesResolver.valueToLiteral(schemaDefault.get(), argument.getType(), getValidationContext().getGraphQLContext(), getValidationContext().getI18n().getLocale());
+        }
```

Comment:
the main fix here is using `argument.getType()` but I rearrange the code a little (moved the `if (expectedType == null)` up so it looks like the change is bigger than it is

Old wrong code

```
schemaDefaultValue = ValuesResolver.valueToLiteral(schemaDefault.get(), expectedType, getValidationContext().getGraphQLContext(), getValidationContext().getI18n().getLocale());
```

new right code

```
schemaDefaultValue = ValuesResolver.valueToLiteral(schemaDefault.get(), argument.getType(), getValidationContext().getGraphQLContext(), getValidationContext().getI18n().getLocale());
```